pkg/webhook/mutating/server.go

```diff
@@ -0,0 +1,102 @@
+package mutating
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+
+	admissionregistrationv1beta1 "k8s.io/api/admissionregistration/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/runtime/inject"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+const (
+	webhookName      = "nmstate"
+	RetrievalTimeout = 120
+)
+
+type server struct {
+	mgr           manager.Manager
+	webhookServer *webhook.Server
+	log           logr.Logger
+}
+
+// Add creates a new Conditions Mutating Webhook and adds it to the Manager. The Manager will set fields on the Webhook
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	return add(mgr, newServer(mgr))
+}
+
+func newServer(mgr manager.Manager) *server {
+	s := &server{
+		webhookServer: &webhook.Server{
+			Port:    8443,
+			CertDir: "/etc/webhook/certs/",
+		},
+		mgr: mgr,
+		log: logf.Log.WithName("webhook/mutating/server"),
+	}
+	s.webhookServer.Register("/mutating", resetConditionsHook())
+	return s
+}
+
+// add adds a new Webhook to mgr with r as the webhook.Server
+func add(mgr manager.Manager, s *server) error {
+	mgr.Add(s)
+	return nil
+}
+
+func (s *server) updateCABundle() error {
+	s.log.Info("Updating CA bundle for webhook")
+	mutatingWebHook := admissionregistrationv1beta1.MutatingWebhookConfiguration{}
+
+	// Do some polling to wait for manifest to be deployed
+	err := wait.PollImmediate(1*time.Second, RetrievalTimeout, func() (bool, error) {
+		webhookKey := types.NamespacedName{Name: webhookName}
+		err := s.mgr.GetClient().Get(context.TODO(), webhookKey, &mutatingWebHook)
+		if err != nil {
+			return false, err
+		}
+		return true, nil
+	})
+	if err != nil {
+		return errors.Wrap(err, "failed retrieving mutationg webhook "+webhookName)
+	}
+	// Retrieve cluster CA bundle and encode to base 64
+	authenticationConfig := corev1.ConfigMap{}
```

Comment:
We should first check if the bundle is already there. It may have been created by another pod.